File: core/FindRightInterval/FindRightInterval.py
from typing import List


def findRightInterval(intervals: List[List[int]]) -> List[int]:
    for i, interval in enumerate(intervals):
        interval.append(i)
    intervals.sort()

    def search(left, right, target):
        while left < right:
            mid = (left + right) // 2
            if intervals[mid][0] == target:
                return intervals[mid][2]
            elif intervals[mid][0] > target:
                right = mid
            else:
                left = mid + 1
        return intervals[left][2] if intervals[left][0] >= target else -1

    length = len(intervals)
    ans = [-1] * length
    for index in range(length):
        ans[intervals[index][2]] = search(0, length - 1, intervals[index][1])
    return ans


if __name__ == '__main__':
    print(findRightInterval([[3,4],[2,3],[1,2]]))

File: core/FindRightInterval/FindRightInterval.java
package com.dbc;

import javafx.util.Pair;

import java.util.Arrays;
import java.util.Comparator;

public class FindRightInterval {
    private int search(int left, int right, int target, int[][] intervals) {
        while (left < right) {
            int mid = (left + right) / 2;
            if (intervals[mid][0] >= target) right = mid;
            else left = mid + 1;
        }
        return intervals[left][0] >= target ? intervals[left][1] : -1;
    }

    public int[] findRightInterval(int[][] intervals) {
        int[][] startIntervals = new int[intervals.length][2];
        for (int i = 0; i < intervals.length; i++) {
            startIntervals[i][0] = intervals[i][0];
            startIntervals[i][1] = i;
        }
        Arrays.sort(startIntervals, Comparator.comparingInt(a -> a[0]));
        int[] ans = new int[intervals.length];
        for (int i = 0; i < intervals.length; i++) {
            ans[i] = search(0, intervals.length - 1, intervals[i][1], startIntervals);
        }
        return ans;
    }
}
